services/recovery_method.js

```diff
@@ -1,69 +1,215 @@
+const stampit = require('@stamp/it');
+
+const { RECOVERY_METHOD_KINDS } = require('../constants');
+const {
+    createRecoveryMethod,
+    deleteRecoveryMethod,
+    getRecoveryMethodByIdentity,
+    listRecoveryMethodsByAccountId,
+    updateRecoveryMethod,
+} = require('../db/methods/recovery_method');
+const { USE_DYNAMODB } = require('../features');
 const SequelizeRecoveryMethods = require('./sequelize/recovery_method');
 
 const TWO_FACTOR_REQUEST_DURATION_MS = 30 * 60000;
 
-const RecoveryMethodService = {
-    createRecoveryMethod({ accountId, detail, kind, publicKey, requestId, securityCode }) {
-        return SequelizeRecoveryMethods.createRecoveryMethod({
-            accountId,
-            detail,
-            kind,
-            publicKey,
-            requestId,
-            securityCode,
-        });
+const RecoveryMethodService = stampit({
+    props: {
+        db: {
+            createRecoveryMethod,
+            deleteRecoveryMethod,
+            getRecoveryMethodByIdentity,
+            listRecoveryMethodsByAccountId,
+            updateRecoveryMethod,
+        },
+        sequelize: SequelizeRecoveryMethods,
     },
+    methods: {
+        createRecoveryMethod({ accountId, detail, kind, publicKey, requestId, securityCode }) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.createRecoveryMethod({
+                    accountId,
+                    detail,
+                    kind,
+                    publicKey,
+                    requestId,
+                    securityCode,
+                });
+            }
+            return this.db.createRecoveryMethod({
+                accountId,
+                detail,
+                kind,
+                publicKey,
+                requestId,
+                securityCode,
+            });
+        },
 
-    deleteOtherRecoveryMethods({ accountId, detail }) {
-        return SequelizeRecoveryMethods.deleteOtherRecoveryMethods({ accountId, detail });
-    },
+        deleteOtherRecoveryMethods({ accountId, detail }) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.deleteOtherRecoveryMethods({ accountId, detail });
+            }
+            return this.db.listRecoveryMethodsByAccountId(accountId)
+                .filter((recoveryMethod) => recoveryMethod.detail !== detail)
+                .map((recoveryMethod) => deleteRecoveryMethod(recoveryMethod));
+        },
 
-    deleteRecoveryMethod({ accountId, kind, publicKey }) {
-        return SequelizeRecoveryMethods.deleteRecoveryMethod({ accountId, kind, publicKey });
-    },
+        deleteRecoveryMethod({ accountId, kind, publicKey }) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.deleteRecoveryMethod({ accountId, kind, publicKey });
+            }
+            return this.db.listRecoveryMethodsByAccountId(accountId)
```

Comment:
Why do we need to list all recovery methods to delete a single method? Is it because we can't construct the `compositeKey` with the present information?

Reply:
Yup exactly, the existing calls only provide those parameters so I've tried to reproduce the same behavior in DynamoDB. It does deviate from the Postgres implementation in that it deletes _all_ methods with those `kind` + `publicKey` values as opposed to arbitrarily taking the first instance. However I believe the only side effect here is that some unused/unseen duplicates may be deleted.